README.md
# NAME

Geo::Region - Geographical regions and groupings using UN M.49 and CLDR data

# VERSION

This document describes Geo::Region for Perl 6, built with Unicode CLDR v26.

# SYNOPSIS

```perl6
use Geo::Region;
use Geo::Region::Enum;

$amer = Geo::Region.new(include => Region::Americas);

$emea = Geo::Region.new(
    include => (Region::Europe, Region::WesternAsia, Region::Africa)
);

$apac = Geo::Region.new(
    include => (Region::Asia, Region::Oceania),
    exclude => Region::WesternAsia,
);

if $amer.contains($country) {
    # country is in the Americas (US, MX, BR, etc.)
}
elsif $emea.contains($country) {
    # country is in Europe, the Middle East, and Africa (FR, SA, ZW, etc.)
}
elsif $apac.contains($country) {
    # country is in Asia-Pacific (JP, TH, AU, etc.)
}
```

# DESCRIPTION

This class is used to create geographical regions and groupings of subregions
and countries. Default regional groupings are provided using the [Unicode CLDR
v26 Territory
Containment](http://unicode.org/cldr/charts/26/supplemental/territory_containment_un_m_49.html)
data, which is an extension of the United Nations [UN
M.49 (Rev.3)](http://unstats.un.org/unsd/methods/m49/m49regin.htm) standard.

## Regions

Regions and subregions are represented with UN M.49 region codes, such as **419**
for Latin America and **035** for Southeast Asia. Either the official format
using a three-digit `0`-padded string like `'035'` or an integer like `35`
may be used with this class. Note when using the `0`-padded format that it must
be quoted as a string so as not to be treated as on octal literal. The CLDR also
adds two additional two-letter region codes which are supported: **EU** for the
European Union and **QO** for Outlying Oceania. These region codes are all
available as enumerations in [Geo::Region::Enum](lib/Geo/Region/Enum.pm).

## Countries

Countries and territories are represented with ISO 3166-1 alpha-2 country codes,
such as **JP** for Japan and **AQ** for Antarctica, and are case insensitive.
Unlike with region codes, the three-digit forms of country codes are not
currently supported, nor are three-letter codes. The deprecated code **UK** for
the United Kingdom is supported as an alias of the official code **GB**.

## Constructor

The `new` class method is used to construct a Geo::Region object along with the
`include` argument and optional `exclude` argument.

- `include`

    Accepts either a single region code or an array reference of region or country
    codes to be included in the resulting custom region.

    ```perl6
    # countries in the European Union (EU)
    Geo::Region.new(include => Region::EuropeanUnion)

    # countries in Asia (142) plus Russia (RU)
    Geo::Region.new(include => (Region::Asia, 'RU'))
    ```

- `exclude`

    Accepts values in the same format as `include`. This can be used to exclude
    countries or subregions from a region.

    ```perl6
    # countries in Europe (150) which are not in the European Union (EU)
    Geo::Region.new(
        include => Region::Europe,
        exclude => Region::EuropeanUnion,
    )
    ```

## Methods

- `contains`

    Given a country or region code, determines if the region represented by the
    Geo::Region instance contains it.

    ```perl6
    if $region.contains($country) {
    ```

- `is-within`

    Given a region code, determines if all the countries and regions represented by
    the Geo::Region instance are within it.

    ```perl6
    if $subregion.is-within($region) {
    ```

- `countries`

    Returns a list of country codes of the countries within the region represented
    by the Geo::Region instance.

    ```perl6
    for $region.countries -> $country {
    ```

# SEE ALSO

- [Geo::Region::Enum](lib/Geo/Region/Enum.pm) — Enumerations for UN M.49 and CLDR region codes
- [Unicode CLDR: UN M.49 Territory
Containment](http://unicode.org/cldr/charts/26/supplemental/territory_containment_un_m_49.html)
- [United Nations: UN M.49 Standard Country, Area, & Region
Codes](http://unstats.un.org/unsd/methods/m49/m49regin.htm)
- [Geo::Region](https://metacpan.org/pod/Geo::Region) for Perl 5

# AUTHOR

Nick Patch <patch@cpan.org>

# COPYRIGHT AND LICENSE

© 2014 Nick Patch

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl 6 itself.


Changes
Revision history for Perl module Geo::Region

0.05 2014-12-05
 - Added Geo::Region::Constant to provide named constants for all supported
   region codes

0.04 2014-12-04
 - Added a warning when using the deprecated `region` constructor argument
   instead of `include`

0.03 2014-11-11
 - Added country code `UK` as a deprecated alias of `GB` (United Kingdom)
 - Changed deprecated region `QU` from being within `150` (Europe) to being a
   deprecated alias of `EU` (European Union), which is also within `150`

0.02 2014-11-09
 - The `region` param to the constructor has been renamed `include` while the
   deprecated alias `region` has been added for backward compatibility
 - Passing the `include` key is now optional when the value is passed as the
   first or only argument to the constructor

0.01 2014-10-10
 - First official release to CPAN
 - Improved docs

0.00_01 2014-10-07
 - First development release to CPAN


t/lib/TestsFor/Geo/Region.pm
package TestsFor::Geo::Region;

use v5.8.1;
use utf8;
use parent qw( Test::Class );
use Test::More;
use Geo::Region;

sub setup :Test(setup) {
    shift->{region} = Geo::Region->new('029');
}

sub object :Test(1) {
    my $region = shift->{region};
    isa_ok $region, 'Geo::Region';
}

sub methods :Test(11) {
    my $region = shift->{region};
    can_ok $region, qw( contains is_within countries )
        or return 'public methods not found';
    ok $region->contains('029'),   'contains itself (029)';
    ok $region->contains('PR'),    'contains Puerto Rico (PR)';
    ok !$region->contains('US'),   'does not contain the United States (US)';
    ok $region->is_within('029'),  'is within itself (029)';
    ok $region->is_within('419'),  'is within Latin America (419)';
    ok $region->is_within('003'),  'is within North America (003)';
    ok $region->is_within('019'),  'is within the Americas (019)';
    ok $region->is_within('001'),  'is within the World (001)';
    ok !$region->is_within('150'), 'is not within Europe (150)';
    is_deeply [$region->countries], [qw(
        AG AI AW BB BL BQ BS CU CW DM DO GD GP HT
        JM KN KY LC MF MQ MS PR SX TC TT VC VG VI
    )], 'countries within region';
}

1;


Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name   => 'Geo::Region',
    license       => 'Perl_5',
    create_readme => 1,
    configure_requires => {
        'Module::Build'     => '0.4004',
        'Software::License' => 0,
    },
    build_requires => {
        'Pod::Man' => '2.04',
    },
    test_requires => {
        'open'              => 0,
        'parent'            => 0,
        'strict'            => 0,
        'warnings'          => 0,
        'English'           => 0,
        'Test::Most'        => '0.22',
    },
    requires => {
        'perl'             => '5.8.1',
        'namespace::clean' => 0,
        'parent'           => 0,
        'utf8'             => 0,
        'Carp'             => 0,
        'Exporter'         => 0,
        'List::Util'       => '1.35',
        'Moo'              => '1.001000',
        'Scalar::Util'     => '1.10',
    },
    meta_merge => {
        resources => {
            repository => 'git://github.com/patch/geo-region-pm5.git',
            homepage   => 'https://github.com/patch/geo-region-pm5',
            bugtracker => 'https://github.com/patch/geo-region-pm5/issues',
        },
    },
);

$builder->create_build_script;


lib/Geo/Region/Constant.pm
module Geo::Region::Constant;

constant AFRICA             is export = '002';
constant AMERICAS           is export = '019';
constant ASIA               is export = '142';
constant AUSTRALASIA        is export = '053';
constant CARIBBEAN          is export = '029';
constant CENTRAL_AMERICA    is export = '013';
constant CENTRAL_ASIA       is export = '143';
constant EASTERN_AFRICA     is export = '014';
constant EASTERN_ASIA       is export = '030';
constant EASTERN_EUROPE     is export = '151';
constant EUROPE             is export = '150';
constant EUROPEAN_UNION     is export = 'EU';
constant LATIN_AMERICA      is export = '419';
constant MELANESIA          is export = '054';
constant MICRONESIAN_REGION is export = '057';
constant MIDDLE_AFRICA      is export = '017';
constant NORTH_AMERICA      is export = '003';
constant NORTHERN_AFRICA    is export = '015';
constant NORTHERN_AMERICA   is export = '021';
constant NORTHERN_EUROPE    is export = '154';
constant OCEANIA            is export = '009';
constant OUTLYING_OCEANIA   is export = 'QO';
constant POLYNESIA          is export = '061';
constant SOUTH_AMERICA      is export = '005';
constant SOUTHEAST_ASIA     is export = '035';
constant SOUTHERN_AFRICA    is export = '018';
constant SOUTHERN_ASIA      is export = '034';
constant SOUTHERN_EUROPE    is export = '039';
constant WESTERN_AFRICA     is export = '011';
constant WESTERN_ASIA       is export = '145';
constant WESTERN_EUROPE     is export = '155';
constant WORLD              is export = '001';

=begin pod

=head1 NAME

Geo::Region::Constant - Constants for UN M.49 and CLDR region codes

=head1 VERSION

This document describes Geo::Region::Constant v0.05, built with Unicode CLDR v26.

=head1 SYNOPSIS

    use Geo::Region::Constant;

=head1 DESCRIPTION

Exportable constants for UN M.49 region codes, including CLDR extensions,
designed for use with L<Geo::Region>.

=head2 Constants

    WORLD                   001
    • AFRICA                002
      ◦ EASTERN_AFRICA      014
      ◦ MIDDLE_AFRICA       017
      ◦ NORTHERN_AFRICA     015
      ◦ SOUTHERN_AFRICA     018
      ◦ WESTERN_AFRICA      011
    • AMERICAS              019
      ◦ CARIBBEAN           029
      ◦ CENTRAL_AMERICA     013
      ◦ LATIN_AMERICA       419
      ◦ NORTH_AMERICA       003
      ◦ NORTHERN_AMERICA    021
      ◦ SOUTH_AMERICA       005
    • ASIA                  142
      ◦ CENTRAL_ASIA        143
      ◦ EASTERN_ASIA        030
      ◦ SOUTHEAST_ASIA      035
      ◦ SOUTHERN_ASIA       034
      ◦ WESTERN_ASIA        145
    • EUROPE                150
      ◦ EASTERN_EUROPE      151
      ◦ EUROPEAN_UNION      EU
      ◦ NORTHERN_EUROPE     154
      ◦ SOUTHERN_EUROPE     039
      ◦ WESTERN_EUROPE      155
    • OCEANIA               009
      ◦ AUSTRALASIA         053
      ◦ MELANESIA           054
      ◦ MICRONESIAN_REGION  057
      ◦ OUTLYING_OCEANIA    QO
      ◦ POLYNESIA           061

=head1 SEE ALSO

=item L<Geo::Region> — Geographical regions and groupings using UN M.49 and CLDR
data
=item L<Unicode CLDR: UN M.49 Territory
Containment|http://unicode.org/cldr/charts/26/supplemental/territory_containment_un_m_49.html>
=item L<United Nations: UN M.49 Standard Country, Area, & Region
Codes|http://unstats.un.org/unsd/methods/m49/m49regin.htm>
=item L<Geo::Region::Constant|https://metacpan.org/pod/Geo::Region::Constant>
for Perl 5

=head1 AUTHOR

Nick Patch <patch@cpan.org>

=head1 COPYRIGHT AND LICENSE

© 2014 Nick Patch

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl 6 itself.

=end pod


lib/Geo/Region.pm
class Geo::Region;

has @!includes;
has @!excludes;
has $!children;
has $!parents;
has $!countries;

submethod BUILD (:$include = (), :$exclude = ()) {
    @!includes = $include.map: { coerce_region($_) };
    @!excludes = $exclude.map: { coerce_region($_) };
}

my %children_of = (
    # regions of subregions
    '001' => <002 009 019 142 150>,
    '002' => <011 014 015 017 018>,
    '003' => <013 021 029>,
    '009' => <053 054 057 061 QO>,
    '019' => <003 005 013 021 029 419>,
    '142' => <030 034 035 143 145>,
    '150' => <039 151 154 155 EU>,
    '419' => <005 013 029>,
    # regions of countries and territories
    '005' => <AR BO BR CL CO EC FK GF GY PE PY SR UY VE>,
    '011' => <BF BJ CI CV GH GM GN GW LR ML MR NE NG SH SL SN TG>,
    '013' => <BZ CR GT HN MX NI PA SV>,
    '014' => <BI DJ ER ET KE KM MG MU MW MZ RE RW SC SO TZ UG YT ZM ZW>,
    '015' => <DZ EA EG EH IC LY MA SD SS TN>,
    '017' => <AO CD CF CG CM GA GQ ST TD ZR>,
    '018' => <BW LS NA SZ ZA>,
    '021' => <BM CA GL PM US>,
    '029' => <AG AI AN AW BB BL BQ BS CU CW DM DO GD GP HT JM KN KY LC MF MQ MS PR SX TC TT VC VG VI>,
    '030' => <CN HK JP KP KR MN MO TW>,
    '034' => <AF BD BT IN IR LK MV NP PK>,
    '035' => <BN BU ID KH LA MM MY PH SG TH TL TP VN>,
    '039' => <AD AL BA CS ES GI GR HR IT ME MK MT PT RS SI SM VA XK YU>,
    '053' => <AU NF NZ>,
    '054' => <FJ NC PG SB VU>,
    '057' => <FM GU KI MH MP NR PW>,
    '061' => <AS CK NU PF PN TK TO TV WF WS>,
    '143' => <KG KZ TJ TM UZ>,
    '145' => <AE AM AZ BH CY GE IL IQ JO KW LB NT OM PS QA SA SY TR YD YE>,
    '151' => <BG BY CZ HU MD PL RO RU SK SU UA>,
    '154' => <AX DK EE FI FO GB GG IE IM IS JE LT LV NO SE SJ>,
    '155' => <AT BE CH DD DE FR FX LI LU MC NL>,
    'EU'  => <AT BE BG CY CZ DE DK EE ES FI FR GB GR HR HU IE IT LT LU LV MT NL PL PT RO SE SI SK>,
    'QO'  => <AC AQ BV CC CP CX DG GS HM IO TA TF UM>,
);

# codes excluded from country list due to being deprecated or grouping container
my $noncountries = set <
    AN BU CS DD FX NT SU TP YD YU ZR
    EU QO
>;

# deprecated aliases
my %alias_of = :QU<EU>, :UK<GB>;

sub coerce_region ($region) {
    return $region.fmt('%03d')
        if $region ~~ /^ <[0..9]>+ $/;

    return %alias_of{$region.uc}
        // $region.uc;
}

method !children () {
    $!children //= do {
        my sub build_children (@regions) {
            @regions.map: {
                $^region,
                %children_of{$^region}:exists
                    ?? build_children(%children_of{$^region})
                    !! ()
            }
        }

        my $excludes = build_children(@!excludes).Set;
        build_children(@!includes).grep({ !$excludes{$_} }).Set;
    };

    return $!children;
}

method !parents () {
    $!parents //= do {
        my sub build_parents (@regions) {
            @regions.map: -> $region {
                $region,
                build_parents(%children_of.keys.grep: {
                    %children_of{$_}.any eq $region
                })
            }
        }

        my %count;
        build_parents(@!includes).grep({
            ++%count{$_} == @!includes
        }).Set;
    };

    return $!parents;
}

method contains ($region) {
    return self!children{ coerce_region($region) };
}

method is-within ($region) {
    return self!parents{ coerce_region($region) };
}

method countries () {
    $!countries //= self!children.keys.grep({
        /<[A..Z]>/ && !$noncountries{$_}
    }).sort;

    return $!countries.values;
}

=begin pod

=head1 NAME

Geo::Region - Geographical regions and groupings using UN M.49 and CLDR data

=head1 VERSION

This document describes Geo::Region for Perl 6, built with Unicode CLDR v26.

=head1 SYNOPSIS

    use Geo::Region;
    use Geo::Region::Enum;

    $amer = Geo::Region.new(include => Region::Americas);

    $emea = Geo::Region.new(
        include => (Region::Europe, Region::WesternAsia, Region::Africa)
    );

    $apac = Geo::Region.new(
        include => (Region::Asia, Region::Oceania),
        exclude => Region::WesternAsia,
    );

    if $amer.contains($country) {
        # country is in the Americas (US, MX, BR, etc.)
    }
    elsif $emea.contains($country) {
        # country is in Europe, the Middle East, and Africa (FR, SA, ZW, etc.)
    }
    elsif $apac.contains($country) {
        # country is in Asia-Pacific (JP, TH, AU, etc.)
    }

=head1 DESCRIPTION

This class is used to create geographical regions and groupings of subregions
and countries. Default regional groupings are provided using the L<Unicode CLDR
v26 Territory
Containment|http://unicode.org/cldr/charts/26/supplemental/territory_containment_un_m_49.html>
data, which is an extension of the United Nations L<UN
M.49 (Rev.3)|http://unstats.un.org/unsd/methods/m49/m49regin.htm> standard.

=head2 Regions

Regions and subregions are represented with UN M.49 region codes, such as B<419>
for Latin America and B<035> for Southeast Asia. Either the official format
using a three-digit C<0>-padded string like C<'035'> or an integer like C<35>
may be used with this class. Note when using the C<0>-padded format that it must
be quoted as a string so as not to be treated as on octal literal. The CLDR also
adds two additional two-letter region codes which are supported: B<EU> for the
European Union and B<QO> for Outlying Oceania. These region codes are all
available as enumerations in L<Geo::Region::Enum>.

=head2 Countries

Countries and territories are represented with ISO 3166-1 alpha-2 country codes,
such as B<JP> for Japan and B<AQ> for Antarctica, and are case insensitive.
Unlike with region codes, the three-digit forms of country codes are not
currently supported, nor are three-letter codes. The deprecated code B<UK> for
the United Kingdom is supported as an alias of the official code B<GB>.

=head2 Constructor

The C<new> class method is used to construct a Geo::Region object along with the
C<include> argument and optional C<exclude> argument.

=begin item
C<include>

Accepts either a single region code or an array reference of region or country
codes to be included in the resulting custom region.

    # countries in the European Union (EU)
    Geo::Region.new(include => Region::EuropeanUnion)

    # countries in Asia (142) plus Russia (RU)
    Geo::Region.new(include => (Region::Asia, Country::Russia))

=end item

=begin item
C<exclude>

Accepts values in the same format as C<include>. This can be used to exclude
countries or subregions from a region.

    # countries in Europe (150) which are not in the European Union (EU)
    Geo::Region.new(
        include => Region::Europe,
        exclude => Region::EuropeanUnion,
    )

=end item

=head2 Methods

=begin item
C<contains>

Given a country or region code, determines if the region represented by the
Geo::Region instance contains it.

    if $region.contains($country) {

=end item

=begin item
C<is-within>

Given a region code, determines if all the countries and regions represented by
the Geo::Region instance are within it.

    if $subregion.is-within($region) {

=end item

=begin item
C<countries>

Returns a list of country codes of the countries within the region represented
by the Geo::Region instance.

    for $region.countries -> $country {

=end item

=head1 SEE ALSO

=item L<Geo::Region::Enum> — Enumerations for UN M.49 and CLDR region codes
=item L<Unicode CLDR: UN M.49 Territory
Containment|http://unicode.org/cldr/charts/26/supplemental/territory_containment_un_m_49.html>
=item L<United Nations: UN M.49 Standard Country, Area, & Region
Codes|http://unstats.un.org/unsd/methods/m49/m49regin.htm>
=item L<Geo::Region|https://metacpan.org/pod/Geo::Region> for Perl 5

=head1 AUTHOR

Nick Patch <patch@cpan.org>

=head1 COPYRIGHT AND LICENSE

© 2014 Nick Patch

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl 6 itself.

=end pod
